File: p3/naive_solution.py
n, m = input().split(' ')
n = int(n); m = int(m)

occ_bits = [0 for i in range(n)]

# Process all m queries
for i in range(m):

    # Parse the query
    query_type, index = input().split(' ')
    query_type = int(query_type); index = int(index)

    if query_type == 1:
        print(occ_bits[index])

    elif query_type == 2:
        occ_bits[index] = 1 - occ_bits[index]
    
    else:
        sum = 0
        for j in range(index + 1): 
            sum += occ_bits[j]; sum = sum % 2
        
        print(sum)

File: p3/NaiveSolution.java
import java.io.*;

public class NaiveSolution {
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));

        String[] inputs = bufferedReader.readLine().replaceAll("\\s+$", "").split(" ");

        int n = Integer.parseInt(inputs[0]);
        int m = Integer.parseInt(inputs[1]);

        // Set up our data structure, which will consist of an array storing the occupation bits
        // and a Fenwick tree (see the solution for more details on what the Fenwick tree does)
        int[] occBits = new int[n];

        // Process all m queries
        for (int i = 0; i < m; i++) {
            int type; int index;

            // Parse the query
            String[] line = bufferedReader.readLine().replaceAll("\\s+$", "").split(" ");
            type = Integer.parseInt(line[0]); index = Integer.parseInt(line[1]);

            switch(type) {
                case 1:
                    System.out.println(occBits[index]);
                    break;

                case 2:
                    occBits[index] = 1 - occBits[index];
                    break;

                case 3:
                    int sum = 0;
                    for (int j = 0; j <= index; j++) sum = (sum + occBits[j]) % 2;
                
                    System.out.println(sum);
            }
        }
    }
}